src/geopackage/wkb.cpp

```diff
@@ -0,0 +1,339 @@
+        for (auto& p : inner) {
+            polygon.inners().at(i).push_back(p);
+        }
+    }
+
+    return polygon;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::multipoint_t wkb::read_multipoint(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    multipoint_t mp;
+    mp.resize(count);
+
+    for (auto& point : mp) {
+        const byte_t new_order = buffer[index];
+        index++;
+
+        uint32_t type;
+        utils::copy_from(buffer, index, type, new_order);
+        throw_if_not_type(type, wkb_geom_t::point);
+
+        point = read_point(buffer, index, new_order);
+    }
+
+    return mp;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::multilinestring_t wkb::read_multilinestring(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    multilinestring_t ml;
+    ml.resize(count);
+    for (auto& line : ml) {
+        const byte_t new_order = buffer[index];
+        index++;
+
+        uint32_t type;
+        utils::copy_from(buffer, index, type, new_order);
+        throw_if_not_type(type, wkb_geom_t::linestring);
+
+        line = read_linestring(buffer, index, new_order);
+    }
+
+    return ml;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::multipolygon_t wkb::read_multipolygon(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    multipolygon_t mpl;
+    mpl.resize(count);
+    for (auto& polygon : mpl) {
+        const byte_t new_order = buffer[index];
+        index++;
+
+        uint32_t type;
+        utils::copy_from(buffer, index, type, new_order);
+        throw_if_not_type(type, wkb_geom_t::polygon);
+
+        polygon = read_polygon(buffer, index, new_order);
+    }
+
+    return mpl;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::geometry wkb::read(const byte_vector& buffer)
+{
+    if (buffer.size() < 5) {
+        throw std::runtime_error("buffer reached end before encountering WKB");
+    }
+
+    int index = 0;
+    const byte_t order = buffer[index];
+    index++;
+
+    uint32_t type;
+    utils::copy_from(buffer, index, type, order);
+
+    switch(type) {
+        case 1:
+            return read_point(buffer, index, order);
+        case 2:
+            return read_linestring(buffer, index, order);
+        case 3:
+            return read_polygon(buffer, index, order);
+        case 4:
+            return read_multipoint(buffer, index, order);
+        case 5:
+            return read_multilinestring(buffer, index, order);
+        case 6:
+            return read_multipolygon(buffer, index, order);
```

Comment:
Please use the named enumeration values from `wkb_geom_t`

Reply:
Implemented in 894e8ab.